include/sys/uart.h

```diff
@@ -0,0 +1,85 @@
+#ifndef _SYS_UART_H_
+#define _SYS_UART_H_
+
+#include <sys/cdefs.h>
+#include <sys/device.h>
+#include <sys/interrupt.h>
+#include <sys/tty.h>
+
+typedef uint8_t (*uart_getc_t)(void *state);
+typedef bool (*uart_rx_ready_t)(void *state);
+
+typedef void (*uart_putc_t)(void *state, uint8_t byte);
+typedef bool (*uart_tx_ready_t)(void *state);
+typedef void (*uart_tx_enable_t)(void *state);
+typedef void (*uart_tx_disable_t)(void *state);
+
+typedef struct uart_methods {
+  uart_getc_t getc;
+  uart_rx_ready_t rx_ready;
+  uart_putc_t putc;
+  uart_tx_ready_t tx_ready;
+  uart_tx_enable_t tx_enable;
+  uart_tx_disable_t tx_disable;
+} uart_methods_t;
+
+static inline uart_methods_t *uart_methods(device_t *dev) {
+  return (uart_methods_t *)(dev)->driver->interfaces[DIF_UART];
+}
+
+typedef struct uart_state {
+  spin_t u_lock;
+  ringbuf_t u_rx_buf;
+  ringbuf_t u_tx_buf;
+  tty_t *u_tty;
```

Comment:
I'm wondering whether we really need to store the TTY as part of the UART's state... Couldn't we store it in `tty_thread_t` instead? I feel like that's where it belongs.

Reply:
Ok, I will do that.

Reply:
done